File: ports/zig/src/launch.zig
const std = @import("std");
const config_mod = @import("config.zig");
const sprite = @import("sprite.zig");

const Config = config_mod.Config;

pub const LaunchArgs = struct {
    dry_run: bool = false,
    no_checkpoint: bool = false,
    upload_dirs: std.ArrayList([]const u8) = .empty,
    sprite_name: ?[]const u8 = null,
    plan_file: ?[]const u8 = null,

    pub fn deinit(self: *LaunchArgs, alloc: std.mem.Allocator) void {
        self.upload_dirs.deinit(alloc);
    }
};

fn printMsg(msg: []const u8) void {
    std.fs.File.stdout().writeAll(msg) catch {};
}

fn printErr(msg: []const u8) void {
    std.fs.File.stderr().writeAll(msg) catch {};
}

pub fn parseArgs(alloc: std.mem.Allocator, raw_args: []const []const u8) !LaunchArgs {
    var args = LaunchArgs{};
    var i: usize = 0;
    while (i < raw_args.len) : (i += 1) {
        const arg = raw_args[i];
        if (std.mem.eql(u8, arg, "--dry-run")) {
            args.dry_run = true;
        } else if (std.mem.eql(u8, arg, "--no-checkpoint")) {
            args.no_checkpoint = true;
        } else if (std.mem.eql(u8, arg, "--upload")) {
            i += 1;
            if (i >= raw_args.len) {
                printErr("Error: --upload requires an argument\n");
                return error.InvalidArgs;
            }
            try args.upload_dirs.append(alloc, raw_args[i]);
        } else if (std.mem.eql(u8, arg, "--help") or std.mem.eql(u8, arg, "-h")) {
            printUsage();
            std.process.exit(0);
        } else if (std.mem.startsWith(u8, arg, "--")) {
            const msg = std.fmt.allocPrint(alloc, "Unknown option: {s}\n", .{arg}) catch {
                printErr("Unknown option\n");
                std.process.exit(1);
            };
            defer alloc.free(msg);
            printErr(msg);
            printUsage();
            std.process.exit(1);
        } else {
            // Positional args: sprite-name [plan-file]
            if (args.sprite_name == null) {
                args.sprite_name = arg;
            } else if (args.plan_file == null) {
                args.plan_file = arg;
            } else {
                printErr("Too many positional arguments\n");
                printUsage();
                std.process.exit(1);
            }
        }
    }

    if (args.sprite_name == null) {
        printUsage();
        std.process.exit(1);
    }

    return args;
}

fn printUsage() void {
    printErr(
        \\Usage: sprite-tool launch [options] <sprite-name> [plan-file]
        \\
        \\Options:
        \\  --dry-run              Show what would happen without executing
        \\  --no-checkpoint        Disable auto-checkpointing
        \\  --upload <dir>         Upload a local directory to /home/sprite/<dirname>
        \\                         (repeatable: --upload ./data --upload ./tests)
        \\
        \\Environment variables:
        \\  ENV_FILE               Path to .env file (default: ./.env)
        \\  SPRITE_TOKEN           sprites.dev API token (or SPRITES_TOKEN in .env)
        \\  AGENT                  "opencode" (default) or "claude"
        \\  CLAUDE_AUTH            "subscription" (default) or "apikey"
        \\  MODEL                  Model override
        \\  CHECKPOINT_INTERVAL    Seconds between checkpoints (default: 300 = 5min)
        \\
    );
}

const CheckpointContext = struct {
    alloc: std.mem.Allocator,
    sprite_name: []const u8,
    interval_ns: u64,
    stop: *std.atomic.Value(bool),
};

fn checkpointThread(ctx: CheckpointContext) void {
    while (!ctx.stop.load(.acquire)) {
        std.Thread.sleep(ctx.interval_ns);
        if (ctx.stop.load(.acquire)) break;

        printMsg("[checkpoint] Creating checkpoint...\n");
        const argv = &[_][]const u8{ "sprite", "checkpoint", "create", "-s", ctx.sprite_name };
        var child = std.process.Child.init(argv, ctx.alloc);
        child.stderr_behavior = .Ignore;
        child.stdout_behavior = .Ignore;
        child.stdin_behavior = .Inherit;
        child.spawn() catch {
            printMsg("[checkpoint] Failed to spawn (non-fatal).\n");
            continue;
        };
        const term = child.wait() catch {
            printMsg("[checkpoint] Failed (non-fatal).\n");
            continue;
        };
        switch (term) {
            .Exited => |code| {
                if (code == 0) {
                    printMsg("[checkpoint] Done.\n");
                } else {
                    printMsg("[checkpoint] Failed (non-fatal).\n");
                }
            },
            else => {
                printMsg("[checkpoint] Failed (non-fatal).\n");
            },
        }
    }
}

pub fn execute(alloc: std.mem.Allocator, raw_args: []const []const u8) !void {
    var args = try parseArgs(alloc, raw_args);
    defer args.deinit(alloc);

    const sprite_name = args.sprite_name.?;

    // Determine env file path from ENV_FILE env var or default
    const env_file_path = std.posix.getenv("ENV_FILE") orelse "./.env";

    // Load .env
    var env_map = try config_mod.loadEnvFile(alloc, env_file_path);
    _ = &env_map; // keep alive
    const cfg = config_mod.buildConfig(&env_map);

    // 1. Check/install sprite CLI
    if (!sprite.isSpriteInstalled(alloc)) {
        if (args.dry_run) {
            printMsg("  [dry-run] Would install sprite CLI\n");
        } else {
            printMsg("Installing sprite CLI...\n");
            const install_argv = &[_][]const u8{ "bash", "-c", "curl -fsSL https://sprites.dev/install.sh | sh" };
            _ = try sprite.run(alloc, install_argv);
        }
    }

    // 2. Auth
    if (cfg.sprite_token) |token| {
        printMsg("Authenticating sprite with token...\n");
        if (!args.dry_run) {
            const auth_argv = &[_][]const u8{ "sprite", "auth", "setup", "--token", token };
            _ = try sprite.run(alloc, auth_argv);
        }
    } else {
        printMsg("No SPRITE_TOKEN set. Running interactive login...\n");
        if (!args.dry_run) {
            const login_argv = &[_][]const u8{ "sprite", "login" };
            _ = try sprite.run(alloc, login_argv);
        }
    }

    // 3. Create sprite (or use existing)
    if (args.dry_run) {
        const msg = try std.fmt.allocPrint(alloc, "  [dry-run] Would create or reuse sprite '{s}'\n", .{sprite_name});
        defer alloc.free(msg);
        printMsg(msg);
    } else {
        // Check if sprite already exists
        const ls_output = try sprite.runCaptureQuiet(alloc, &[_][]const u8{ "sprite", "ls" });
        var exists = false;
        if (ls_output) |output| {
            defer alloc.free(output);
            var lines = std.mem.splitScalar(u8, output, '\n');
            while (lines.next()) |line| {
                var words = std.mem.splitAny(u8, line, " \t");
                while (words.next()) |word| {
                    if (std.mem.eql(u8, word, sprite_name)) {
                        exists = true;
                        break;
                    }
                }
                if (exists) break;
            }
        }

        if (exists) {
            const msg = try std.fmt.allocPrint(alloc, "Sprite '{s}' already exists, using it.\n", .{sprite_name});
            defer alloc.free(msg);
            printMsg(msg);
        } else {
            const msg = try std.fmt.allocPrint(alloc, "Creating sprite: {s}\n", .{sprite_name});
            defer alloc.free(msg);
            printMsg(msg);
            const create_argv = &[_][]const u8{ "sprite", "create", "-skip-console", sprite_name };
            _ = try sprite.run(alloc, create_argv);
        }
    }

    // 4. Push .env to sprite
    if (fileExists(env_file_path)) {
        const msg = try std.fmt.allocPrint(alloc, "Pushing {s}...\n", .{env_file_path});
        defer alloc.free(msg);
        printMsg(msg);
        try sprite.pushFile(alloc, sprite_name, env_file_path, "/home/sprite/.env", args.dry_run);
    }

    // 5. Push plan file if provided
    if (args.plan_file) |plan_file| {
        if (fileExists(plan_file)) {
            const msg = try std.fmt.allocPrint(alloc, "Pushing {s}...\n", .{plan_file});
            defer alloc.free(msg);
            printMsg(msg);
            try sprite.pushFile(alloc, sprite_name, plan_file, "/home/sprite/plan.md", args.dry_run);
        }
    }

    // 6. Upload directories
    for (args.upload_dirs.items) |dir| {
        if (dirExists(dir)) {
            const basename = std.fs.path.basename(dir);
            const remote = try std.fmt.allocPrint(alloc, "/home/sprite/{s}", .{basename});
            defer alloc.free(remote);
            const msg = try std.fmt.allocPrint(alloc, "Uploading directory: {s} -> {s}\n", .{ dir, remote });
            defer alloc.free(msg);
            printMsg(msg);
            try sprite.pushDir(alloc, sprite_name, dir, remote, args.dry_run);
        } else {
            const msg = try std.fmt.allocPrint(alloc, "WARNING: --upload dir '{s}' not found, skipping.\n", .{dir});
            defer alloc.free(msg);
            printMsg(msg);
        }
    }

    // 7. Setup git + beads
    printMsg("Initializing git...\n");
    _ = try sprite.sx(alloc, sprite_name, "cd /home/sprite && git init -b main 2>/dev/null || true", args.dry_run);

    printMsg("Installing beads...\n");
    _ = try sprite.sx(alloc, sprite_name, "curl -fsSL https://raw.githubusercontent.com/steveyegge/beads/main/scripts/install.sh | bash", args.dry_run);

    // 8. Install and auth coding agent
    if (std.mem.eql(u8, cfg.agent, "claude")) {
        printMsg("Setting up claude...\n");
        _ = try sprite.sx(alloc, sprite_name, "command -v claude >/dev/null 2>&1 || npm install -g @anthropic-ai/claude-code", args.dry_run);

        if (std.mem.eql(u8, cfg.claude_auth, "subscription")) {
            const home = std.posix.getenv("HOME") orelse "/root";
            const cred_path = try std.fmt.allocPrint(alloc, "{s}/.claude/.credentials.json", .{home});
            defer alloc.free(cred_path);
            if (fileExists(cred_path)) {
                printMsg("Copying claude subscription credentials...\n");
                try sprite.pushFile(alloc, sprite_name, cred_path, "/home/sprite/.claude/.credentials.json", args.dry_run);
                _ = try sprite.sx(alloc, sprite_name, "chmod 600 ~/.claude/.credentials.json", args.dry_run);
            } else {
                printErr("ERROR: ~/.claude/.credentials.json not found\n");
                printErr("Run 'claude' locally first to authenticate, then re-run this script.\n");
                std.process.exit(1);
            }
        } else if (std.mem.eql(u8, cfg.claude_auth, "apikey")) {
            if (cfg.anthropic_api_key) |api_key| {
                printMsg("Setting ANTHROPIC_API_KEY in sprite...\n");
                const cmd = try std.fmt.allocPrint(alloc, "grep -q ANTHROPIC_API_KEY ~/.bashrc 2>/dev/null || echo 'export ANTHROPIC_API_KEY=\"{s}\"' >> ~/.bashrc", .{api_key});
                defer alloc.free(cmd);
                _ = try sprite.sx(alloc, sprite_name, cmd, args.dry_run);
            } else {
                printErr("ERROR: CLAUDE_AUTH=apikey but ANTHROPIC_API_KEY is not set\n");
                std.process.exit(1);
            }
        } else {
            printErr("ERROR: No valid claude auth configured\n");
            std.process.exit(1);
        }
    } else if (std.mem.eql(u8, cfg.agent, "opencode")) {
        printMsg("Setting up opencode...\n");
        _ = try sprite.sx(alloc, sprite_name, "[ -x ~/.opencode/bin/opencode ] || curl -fsSL https://opencode.ai/install | bash", args.dry_run);
        _ = try sprite.sx(alloc, sprite_name, "grep -q 'source.*\\.env' ~/.bashrc 2>/dev/null || echo '[ -f /home/sprite/.env ] && set -a && source /home/sprite/.env && set +a' >> ~/.bashrc", args.dry_run);
    } else {
        const msg = try std.fmt.allocPrint(alloc, "ERROR: Unknown AGENT '{s}'. Use 'claude' or 'opencode'.\n", .{cfg.agent});
        defer alloc.free(msg);
        printErr(msg);
        std.process.exit(1);
    }

    // 9. Launch agent with plan (or open console)
    printMsg("\n==========================================\n");
    {
        const msg = try std.fmt.allocPrint(alloc, "Sprite '{s}' is ready!\nAgent: {s}", .{ sprite_name, cfg.agent });
        defer alloc.free(msg);
        printMsg(msg);
    }
    if (cfg.model) |m| {
        const msg = try std.fmt.allocPrint(alloc, " (model: {s})", .{m});
        defer alloc.free(msg);
        printMsg(msg);
    }
    printMsg("\n");
    if (!args.no_checkpoint) {
        const msg = try std.fmt.allocPrint(alloc, "Checkpointing: every {d}s\n", .{cfg.checkpoint_interval});
        defer alloc.free(msg);
        printMsg(msg);
    }
    printMsg("==========================================\n");

    if (args.dry_run) {
        const msg = try std.fmt.allocPrint(alloc, "\n[dry-run] Would launch {s} with plan. No changes were made.\n", .{cfg.agent});
        defer alloc.free(msg);
        printMsg(msg);
        return;
    }

    if (args.plan_file != null) {
        // Start auto-checkpointing
        var stop_flag = std.atomic.Value(bool).init(false);
        var checkpoint_thread: ?std.Thread = null;

        if (!args.no_checkpoint) {
            const interval_ns = cfg.checkpoint_interval * std.time.ns_per_s;
            const ctx = CheckpointContext{
                .alloc = alloc,
                .sprite_name = sprite_name,
                .interval_ns = interval_ns,
                .stop = &stop_flag,
            };
            checkpoint_thread = try std.Thread.spawn(.{}, checkpointThread, .{ctx});
            const msg = try std.fmt.allocPrint(alloc, "Auto-checkpointing every {d}s\n", .{cfg.checkpoint_interval});
            defer alloc.free(msg);
            printMsg(msg);
        }

        {
            const msg = try std.fmt.allocPrint(alloc, "Launching {s} with plan...\n", .{cfg.agent});
            defer alloc.free(msg);
            printMsg(msg);
        }

        if (std.mem.eql(u8, cfg.agent, "claude")) {
            var cmd_buf: std.ArrayList(u8) = .empty;
            defer cmd_buf.deinit(alloc);
            try cmd_buf.appendSlice(alloc, "cd /home/sprite && claude ");
            if (cfg.model) |m| {
                try cmd_buf.appendSlice(alloc, "--model ");
                try cmd_buf.appendSlice(alloc, m);
                try cmd_buf.append(alloc, ' ');
            }
            try cmd_buf.appendSlice(alloc, "-p 'read plan.md and complete the plan please'");
            _ = try sprite.sx(alloc, sprite_name, cmd_buf.items, false);
        } else if (std.mem.eql(u8, cfg.agent, "opencode")) {
            const oc_model = cfg.model orelse "opencode/big-pickle";
            const cmd = try std.fmt.allocPrint(alloc, "set -a && source /home/sprite/.env 2>/dev/null && set +a && cd /home/sprite && ~/.opencode/bin/opencode run -m {s} 'read plan.md and complete the plan please'", .{oc_model});
            defer alloc.free(cmd);
            _ = try sprite.sx(alloc, sprite_name, cmd, false);
        }

        // Stop checkpointing
        stop_flag.store(true, .release);
        if (checkpoint_thread) |t| {
            t.join();
        }

        // Final checkpoint
        printMsg("Creating final checkpoint...\n");
        const ck_argv = &[_][]const u8{ "sprite", "checkpoint", "create", "-s", sprite_name };
        if (try sprite.run(alloc, ck_argv)) {
            printMsg("Final checkpoint saved.\n");
        } else {
            printMsg("Final checkpoint failed (non-fatal).\n");
        }
    } else {
        printMsg("Opening console...\n");
        const console_argv = &[_][]const u8{ "sprite", "console", "-s", sprite_name };
        _ = try sprite.run(alloc, console_argv);
    }
}

fn fileExists(path: []const u8) bool {
    std.fs.cwd().access(path, .{}) catch return false;
    return true;
}

fn dirExists(path: []const u8) bool {
    var dir = std.fs.cwd().openDir(path, .{}) catch return false;
    dir.close();
    return true;
}

File: ports/lua/lib/launch.lua
-- launch.lua: sprite-tool launch — Create and configure a sprite with coding agent, git, beads.

local config = require("config")
local sprite = require("sprite")

local M = {}

local function usage()
    print([[Usage: sprite-tool launch [options] <sprite-name> [plan-file]

Options:
  --dry-run              Show what would happen without executing
  --no-checkpoint        Disable auto-checkpointing
  --upload <dir>         Upload a local directory to /home/sprite/<dirname>
                         (repeatable: --upload ./data --upload ./tests)

Environment variables:
  ENV_FILE               Path to .env file (default: ./.env)
  SPRITE_TOKEN           sprites.dev API token (or SPRITES_TOKEN in .env)
  AGENT                  "opencode" (default) or "claude"
  CLAUDE_AUTH            "subscription" (default) or "apikey"
  MODEL                  Model override (see below)
  CHECKPOINT_INTERVAL    Seconds between checkpoints (default: 300 = 5min)

Model examples:
  OpenCode: MODEL=opencode/big-pickle  (free, default)
            MODEL=groq/llama-3.3-70b-versatile  (free w/ GROQ_API_KEY)
            MODEL=openai/gpt-4o
            MODEL=google/gemini-2.5-pro
  Claude:   MODEL=opus, MODEL=sonnet, MODEL=haiku

Examples:
  sprite-tool launch my-project plan.md
  sprite-tool launch --upload ./data my-project plan.md
  sprite-tool launch --upload ./data --upload ./tests my-project plan.md
  sprite-tool launch --dry-run my-project plan.md
  MODEL=groq/llama-3.3-70b-versatile sprite-tool launch dev plan.md
  AGENT=claude MODEL=sonnet sprite-tool launch dev plan.md]])
    os.exit(1)
end

--- Check if a file exists.
local function file_exists(path)
    local fh = io.open(path, "r")
    if fh then
        fh:close()
        return true
    end
    return false
end

--- Check if a directory exists.
local function dir_exists(path)
    -- Use test -d via shell; os.execute returns true on success in Lua 5.2+
    local ok = os.execute("test -d " .. config.shell_quote(path))
    return ok == true or ok == 0
end

--- Start background checkpointing using a shell background process.
-- Returns a description string. The background process runs until the
-- parent shell pipeline ends or we explicitly kill it.
local function start_checkpointing(sprite_name, interval)
    local sq = config.shell_quote(sprite_name)
    -- Launch a background subshell that loops: sleep, checkpoint, repeat
    local bg_cmd = string.format(
        [[(while true; do sleep %d; echo "[checkpoint] Creating checkpoint at $(date '+%%H:%%M:%%S')..."; sprite checkpoint create -s %s 2>/dev/null && echo "[checkpoint] Done." || echo "[checkpoint] Failed (non-fatal)."; done) &]],
        interval, sq
    )
    os.execute(bg_cmd)
    print("Auto-checkpointing every " .. interval .. "s (background process)")
end

--- Stop background checkpointing by killing any matching background sleep/checkpoint loops.
local function stop_checkpointing(sprite_name)
    -- Best-effort kill of the background checkpoint loop
    local sq = config.shell_quote(sprite_name)
    os.execute("pkill -f 'sprite checkpoint create -s " .. sprite_name .. "' 2>/dev/null || true")
end

--- Main entry point for the launch subcommand.
function M.run(args)
    -- Load config (parses .env, reads env vars)
    local cfg = config.load()

    -- Parse flags
    local dry_run = false
    local checkpointing = true
    local upload_dirs = {}

    local i = 1
    while i <= #args and args[i]:sub(1, 2) == "--" do
        local flag = args[i]
        if flag == "--dry-run" then
            dry_run = true
            i = i + 1
        elseif flag == "--no-checkpoint" then
            checkpointing = false
            i = i + 1
        elseif flag == "--upload" then
            i = i + 1
            if i > #args then usage() end
            upload_dirs[#upload_dirs + 1] = args[i]
            i = i + 1
        elseif flag == "--help" or flag == "-h" then
            usage()
        else
            io.stderr:write("Unknown option: " .. flag .. "\n")
            usage()
        end
    end

    if i > #args then usage() end

    local sprite_name = args[i]
    i = i + 1
    local plan_file = args[i] -- may be nil

    -- Store into config
    cfg.dry_run = dry_run
    cfg.checkpointing = checkpointing
    cfg.upload_dirs = upload_dirs

    -- ============================================================
    -- 1. Check/install sprite CLI
    -- ============================================================
    local has_sprite = os.execute("command -v sprite >/dev/null 2>&1")
    if not (has_sprite == true or has_sprite == 0) then
        if dry_run then
            print("  [dry-run] Would install sprite CLI")
        else
            print("Installing sprite CLI...")
            os.execute("curl -fsSL https://sprites.dev/install.sh | sh")
            -- Note: PATH update only affects child processes in Lua
        end
    end

    -- ============================================================
    -- 2. Auth sprite (non-interactive if token provided)
    -- ============================================================
    if cfg.sprite_token ~= "" then
        print("Authenticating sprite with token...")
        if not dry_run then
            os.execute("sprite auth setup --token " .. config.shell_quote(cfg.sprite_token))
        end
    else
        print("No SPRITE_TOKEN set. Running interactive login...")
        if not dry_run then
            os.execute("sprite login")
        end
    end

    -- ============================================================
    -- 3. Create sprite (or use existing)
    -- ============================================================
    if dry_run then
        print("  [dry-run] Would create or reuse sprite '" .. sprite_name .. "'")
    else
        local handle = io.popen("sprite ls 2>/dev/null", "r")
        local ls_output = ""
        if handle then
            ls_output = handle:read("*a") or ""
            handle:close()
        end

        -- Check if sprite name appears as a whole word in ls output
        if ls_output:find(sprite_name, 1, true) then
            print("Sprite '" .. sprite_name .. "' already exists, using it.")
        else
            print("Creating sprite: " .. sprite_name)
            os.execute("sprite create -skip-console " .. config.shell_quote(sprite_name))
        end
    end

    -- ============================================================
    -- 4. Push .env to sprite
    -- ============================================================
    if file_exists(cfg.env_file) then
        print("Pushing " .. cfg.env_file .. "...")
        sprite.push_file(sprite_name, cfg.env_file, "/home/sprite/.env", dry_run)
    end

    -- ============================================================
    -- 5. Push plan file if provided
    -- ============================================================
    if plan_file and plan_file ~= "" and file_exists(plan_file) then
        print("Pushing " .. plan_file .. "...")
        sprite.push_file(sprite_name, plan_file, "/home/sprite/plan.md", dry_run)
    end

    -- ============================================================
    -- 6. Upload directories if provided
    -- ============================================================
    for _, dir in ipairs(upload_dirs) do
        if dir_exists(dir) then
            local dirname = sprite.basename(dir)
            print("Uploading directory: " .. dir .. " -> /home/sprite/" .. dirname)
            sprite.push_dir(sprite_name, dir, "/home/sprite/" .. dirname, dry_run)
        else
            print("WARNING: --upload dir '" .. dir .. "' not found, skipping.")
        end
    end

    -- ============================================================
    -- 7. Setup git + beads
    -- ============================================================
    print("Initializing git...")
    sprite.sx(sprite_name, "cd /home/sprite && git init -b main 2>/dev/null || true", dry_run)

    print("Installing beads...")
    sprite.sx(sprite_name, "curl -fsSL https://raw.githubusercontent.com/steveyegge/beads/main/scripts/install.sh | bash", dry_run)

    -- ============================================================
    -- 8. Install and auth coding agent
    -- ============================================================
    local agent = cfg.agent
    local model = cfg.model

    if agent == "claude" then
        print("Setting up claude...")
        sprite.sx(sprite_name, "command -v claude >/dev/null 2>&1 || npm install -g @anthropic-ai/claude-code", dry_run)

        if cfg.claude_auth == "subscription" then
            local home = os.getenv("HOME") or ""
            local creds = home .. "/.claude/.credentials.json"
            if file_exists(creds) then
                print("Copying claude subscription credentials...")
                sprite.push_file(sprite_name, creds, "/home/sprite/.claude/.credentials.json", dry_run)
                sprite.sx(sprite_name, "chmod 600 ~/.claude/.credentials.json", dry_run)
            else
                io.stderr:write("ERROR: ~/.claude/.credentials.json not found\n")
                io.stderr:write("Run 'claude' locally first to authenticate, then re-run this script.\n")
                os.exit(1)
            end
        elseif cfg.claude_auth == "apikey" and cfg.anthropic_api_key ~= "" then
            print("Setting ANTHROPIC_API_KEY in sprite...")
            local api_key = cfg.anthropic_api_key
            sprite.sx(sprite_name,
                "grep -q ANTHROPIC_API_KEY ~/.bashrc 2>/dev/null || echo 'export ANTHROPIC_API_KEY=\"" .. api_key .. "\"' >> ~/.bashrc",
                dry_run)
        else
            io.stderr:write("ERROR: No valid claude auth configured\n")
            io.stderr:write("Set CLAUDE_AUTH=subscription (default) or CLAUDE_AUTH=apikey with ANTHROPIC_API_KEY\n")
            os.exit(1)
        end

    elseif agent == "opencode" then
        print("Setting up opencode...")
        sprite.sx(sprite_name, "[ -x ~/.opencode/bin/opencode ] || curl -fsSL https://opencode.ai/install | bash", dry_run)
        sprite.sx(sprite_name,
            [[grep -q 'source.*\.env' ~/.bashrc 2>/dev/null || echo '[ -f /home/sprite/.env ] && set -a && source /home/sprite/.env && set +a' >> ~/.bashrc]],
            dry_run)
    else
        io.stderr:write("ERROR: Unknown AGENT '" .. agent .. "'. Use 'claude' or 'opencode'.\n")
        os.exit(1)
    end

    -- ============================================================
    -- 9. Launch agent with plan (or open console)
    -- ============================================================
    print("")
    print("==========================================")
    print("Sprite '" .. sprite_name .. "' is ready!")
    local model_info = ""
    if model ~= "" then
        model_info = " (model: " .. model .. ")"
    end
    print("Agent: " .. agent .. model_info)
    if checkpointing then
        print("Checkpointing: every " .. cfg.checkpoint_interval .. "s")
    end
    print("==========================================")

    if dry_run then
        print("")
        print("[dry-run] Would launch " .. agent .. " with plan. No changes were made.")
        return 0
    end

    if plan_file and plan_file ~= "" then
        -- Start auto-checkpointing before agent runs
        if checkpointing then
            start_checkpointing(sprite_name, cfg.checkpoint_interval)
        end

        print("Launching " .. agent .. " with plan...")

        if agent == "claude" then
            local model_flag = ""
            if model ~= "" then
                model_flag = "--model " .. model .. " "
            end
            sprite.sx(sprite_name,
                "cd /home/sprite && claude " .. model_flag .. "-p 'read plan.md and complete the plan please'")

        elseif agent == "opencode" then
            local oc_model = model ~= "" and model or "opencode/big-pickle"
            sprite.sx(sprite_name,
                "set -a && source /home/sprite/.env 2>/dev/null && set +a && cd /home/sprite && ~/.opencode/bin/opencode run -m " .. oc_model .. " 'read plan.md and complete the plan please'")
        end

        -- Final checkpoint after agent completes
        if checkpointing then
            stop_checkpointing(sprite_name)
        end
        print("Creating final checkpoint...")
        local ok = os.execute("sprite checkpoint create -s " .. config.shell_quote(sprite_name) .. " 2>/dev/null")
        if ok == true or ok == 0 then
            print("Final checkpoint saved.")
        else
            print("Final checkpoint failed (non-fatal).")
        end
    else
        print("Opening console...")
        os.execute("sprite console -s " .. config.shell_quote(sprite_name))
    end

    return 0
end

return M
